File: app/src/test/java/com/annushkaproject/programmerscalculator/StandardTests/StandardAdditionAndSubtractionTest.java
package com.annushkaproject.programmerscalculator.StandardTests;

import com.annushkaproject.programmerscalculator.model.CalculationModel;
import com.annushkaproject.programmerscalculator.model.Value;
import com.annushkaproject.programmerscalculator.utils.StandardOperationsUtil;

import org.junit.Test;

import static com.annushkaproject.programmerscalculator.model.Operator.ADD;
import static com.annushkaproject.programmerscalculator.model.Operator.SUBTRACT;
import static org.junit.Assert.assertEquals;

public class StandardAdditionAndSubtractionTest {
    @Test
    public void standardAddition_isCorrect() {
        assertEquals(10, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(5), new Value(5), ADD)), 0);
        assertEquals(-10, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(-5), new Value(5), SUBTRACT)), 0);
        assertEquals(-0.888, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(1.112), new Value(2), SUBTRACT)), 0.000001);
        assertEquals(0, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(0x4000000000000000L), new Value(0x4000000000000000L), SUBTRACT)), 0);
        assertEquals(-9.223372036854776E18, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(-0x8000000000000000L), new Value(1), SUBTRACT)), 0);
        assertEquals(9.223372036854776E18, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(0x7FFFFFFFFFFFFFFFL), new Value(1), ADD)), 0);
        assertEquals(9.223372036854776E18, StandardOperationsUtil.calculateWithData(new CalculationModel(new Value(0x4000000000000000L), new Value(0x4000000000000000L), ADD)), 0);
    }
}

File: app/src/test/java/com/annushkaproject/programmerscalculator/StandardTests/StandardAdditionAndSubtractionTest.kt
package com.annushkaproject.programmerscalculator.StandardTests

import com.annushkaproject.programmerscalculator.model.*
import com.annushkaproject.programmerscalculator.utils.StandardOperationsUtil.calculateResultForTwoSidedOperator
import org.junit.Assert
import org.junit.Test
import java.math.BigDecimal

class StandardAdditionAndSubtractionTest {
    @Test
    fun testAddIwoIntegers() {
        val firstValue = BigDecimal.valueOf(5)
        val secondValue = BigDecimal.valueOf(5)
        val model = CalculationModel(firstValue, secondValue, Operator.ADD)
        Assert.assertEquals(10.0, calculateResultForTwoSidedOperator(model), 0.0)
    }

    @Test
    fun testSubtractIwoIntegers() {
        val firstValue = BigDecimal.valueOf(-5)
        val secondValue = BigDecimal.valueOf(5)
        val model = CalculationModel(firstValue, secondValue, Operator.SUBTRACT)
        Assert.assertEquals(-10.0, calculateResultForTwoSidedOperator(model), 0.0)
    }

    @Test
    fun testSubtractFloatAndDecimal() {
        val firstValue = BigDecimal.valueOf(1.112)
        val secondValue = BigDecimal.valueOf(2)
        val model = CalculationModel(firstValue, secondValue, Operator.SUBTRACT)
        Assert.assertEquals(-0.888, calculateResultForTwoSidedOperator(model), 0.000001)
    }

    @Test
    fun testUnderflow() {
        val firstValue: BigDecimal = BigDecimal.valueOf(--0x8000000000000000L)
        val secondValue = BigDecimal.valueOf(1)
        val model = CalculationModel(firstValue, secondValue, Operator.SUBTRACT)
        Assert.assertEquals(-9.223372036854776E18, calculateResultForTwoSidedOperator(model), 0.0)
    }

    @Test
    fun testOverflow() {
        val firstValue = BigDecimal.valueOf(0x7FFFFFFFFFFFFFFFL)
        val secondValue = BigDecimal.valueOf(1)
        val model = CalculationModel(firstValue, secondValue, Operator.ADD)
        Assert.assertEquals(9.223372036854776E18, calculateResultForTwoSidedOperator(model), 0.0)
    }

    @Test
    fun testOverflowSameValues() {
        val firstValue = BigDecimal.valueOf(0x4000000000000000L)
        val secondValue = BigDecimal.valueOf(0x4000000000000000L)
        val model = CalculationModel(firstValue, secondValue, Operator.ADD)
        Assert.assertEquals(9.223372036854776E18, calculateResultForTwoSidedOperator(model), 0.0)
    }

    @Test
    fun testUnderflowSameValues() {
        val firstValue = BigDecimal.valueOf(0x4000000000000000L)
        val secondValue = BigDecimal.valueOf(0x4000000000000000L)
        val model = CalculationModel(firstValue, secondValue, Operator.SUBTRACT)
        Assert.assertEquals(0.0, calculateResultForTwoSidedOperator(model), 0.0)
    }
}